Separated ApplicationId and identity column of Application model

xperiandri opened the pull request:
Issue #112

Also eliminated dependency of Application primary key type from ASP.NET Identity primary key type.
Added consytaints of 38 characters for columns of Application model used for indexes.
Added XML documentation to modified types.
Updated sample Server project.

kevinchalet commented on samples/Mvc.Server/Startup.cs:
```diff
@@ -135,7 +135,7 @@ public class Startup {
 
                     context.Applications.Add(new Application {
-                        Id = "myClient",
+                        ApplicationId = "myClient",
```
Maybe it would be clearer to name it `ClientId`?

xperiandri replied:
Then maybe we rename Application entity to Client entity? Or do you have some intent of having different names?

kevinchalet replied:
Having both `Id` and `ApplicationId` in the same `Application` entity is confusing. Renaming the second one would make clearer that `ClientId` is not a primary key.

kevinchalet commented on src/OpenIddict.EF/OpenIddictContext.cs:
```diff
@@ -10,29 +10,120 @@
+    /// <typeparam name="TRole">The type of role objects.</typeparam>
+    /// <typeparam name="TKey">The type of the primary key for users and roles.</typeparam>
+    /// <seealso cref="Microsoft.AspNetCore.Identity.EntityFrameworkCore.IdentityDbContext{TUser, TRole, TKey}" />
+    public class OpenIddictContext<TApplication, TApplicationKey, TUser, TRole, TKey> : IdentityDbContext<TUser, TRole, TKey>
```
I'm pretty sure we said "no" to a separate key type. Use TKey instead :smile:

xperiandri replied:
What if I want to use a `Guid` as a key for Identity and a `string` as an `ApplicationId` and a key at the same time? For me there is no reason to add additional column for primary key as `ApplicationId` is perfect option. But as long as `Application` primary key depends on Identity primary key I can't do that!
I implemented compromise variant. See additional `OpenIddictContext<TApplication, TUser, TRole, TKey>` implementation.

kevinchalet replied:
> What if I want to use a Guid as a key for Identity and string as an ApplicationId and key at the same time?

You won't be able to do that, just like Identity doesn't allow you to use different key types for user/role. With the same logic, we'd have to add a `TAuthorizationKey`, a `TScopeKey`, a `TTokenKey` and honestly, it would be a real nightmare.

Don't forget that Identity added new generic arguments recently: https://github.com/aspnet/Identity/pull/798/files#diff-c72544a50eacab8ada8ca53f53daee15R84.

That would be great if we could avoid having 20 or 30 generic arguments :fearful:

xperiandri replied:
I understand you point however `Application` has nothing in common with Identity. Why must it depend on its generic parameter?
And your constraint breaks my primary scenario.

xperiandri replied:
> TAuthorizationKey , a  TScopeKey , a  TTokenKey

I need some explanation. I haven't looked there.

kevinchalet replied:
Similarly to what Identity does, the primary keys of all the OpenIddict entities are of type `TKey`.

kevinchalet replied:
> And your constraint breaks my primary scenario.

It won't if we separate `Id` (the primary key, whose type is `TKey`) and `ApplicationId`/`ClientId`, whose type will necessarily be a string.

kevinchalet replied:
> I need some explanation. I haven't looked there.

We are going to introduce new entities: `Authorization`, `Scope` and `Token`. If you want to introduce a key type specific to `Application`, then the same rule should apply to `Scope` and `Token`.

xperiandri replied:
So you mean that you are going to have relationships with another entites from `Application` entity, right? And for that purpose you want to have the same key type on `Application` and that entities.

kevinchalet replied:
No, I mean that if you want to have a key type specific to the `Application` entity, then someone else will want a key type for `Authorization`, another one for `Scope` and a last one for `Token`.

We can't afford having a key type for each entity.

xperiandri replied:
What are `Authorization`, `Scope` and `Token`? I don't see such entities.
`Application` is a standalone entity with no relationships now. I don't see a relationship to what you say.

xperiandri replied:
Are you going to save `Authorization`, `Scope` and `Token` to database and link them to `Application`?

kevinchalet replied:
These entities are coming as part of my PR. E.g there's a direct relationship between `OpenIddictAuthorization` and `OpenIddictToken`:

``` csharp
/// <summary>
/// Represents an OpenIddict authorization.
/// </summary>
public class OpenIddictAuthorization<TToken, TKey> where TKey : IEquatable<TKey> {
    /// <summary>
    /// Gets or sets the unique identifier
    /// associated with the current authorization.
    /// </summary>
    public virtual TKey Id { get; set; }

    /// <summary>
    /// Gets or sets the space-delimited scopes
    /// associated with the current authorization.
    /// </summary>
    public virtual string Scope { get; set; }

    /// <summary>
    /// Gets or sets the list of tokens
    /// associated with the current authorization.
    /// </summary>
    public virtual IList<TToken> Tokens { get; } = new List<TToken>();

    /// <summary>
    /// Gets or sets the identifier of the user profile
    /// associated with the current authorization.
    /// </summary>
    public virtual TKey UserId { get; set; }
}
```

``` csharp
/// <summary>
/// Represents an OpenIddict token.
/// </summary>
public class OpenIddictToken<TKey> where TKey : IEquatable<TKey> {
    /// <summary>
    /// Gets or sets the identifier of the authorization attached with the current token.
    /// This property may be null if the token was issued without
    /// requiring the user consent or is bound to a client application.
    /// </summary>
    public virtual TKey AuthorizationId { get; set; }

    /// <summary>
    /// Gets or sets the unique identifier
    /// associated with the current token.
    /// </summary>
    public virtual TKey Id { get; set; }

    /// <summary>
    /// Gets or sets the type of the current token.
    /// </summary>
    public virtual string Type { get; set; }

    /// <summary>
    /// Gets or sets the identifier of the user attached with the current token.
    /// This property is null if the token represents a client application.
    /// </summary>
    public virtual TKey UserId { get; set; }
}
```

kevinchalet commented on src/OpenIddict.EF/OpenIddictExtensions.cs:
```diff
@@ -74,6 +75,30 @@ where typeof(DbContext).IsAssignableFrom(argument)
             throw new InvalidOperationException("The type of the database context cannot be automatically inferred.");
         }
 
+        private static Type ResolveApplicationKeyType([NotNull] OpenIddictConfiguration configuration) {
```
FYI, all this reflection sorcery is going to disappear with my experimental PR.

kevinchalet commented on src/OpenIddict.Models/Application.cs:
```diff
@@ -5,16 +5,82 @@
+    /// </remarks>
+    public class Application : Application<string>
+    {
+        public override string ApplicationId
```
Looks like rather unusual. Is this a pattern Identity uses somewhere? I'm concerned about introducing something totally new here.

xperiandri replied:
Actually they allow you to override `Id` on `IdentityUser` for example.

kevinchalet commented:
Thanks for the PR! Minor a few remarks, it looks nice.

That said, it will directly conflict with my own changes. I'll try to merge my PR soon so we can progress with yours.

xperiandri commented:
OK
At least now I can build my current web application from my fork of sources and migrations work.
Because before I couldn't use Identity keys of `Guid` and ApplicationIds of `string`.

kevinchalet commented:
> Because before I couldn't use Identity keys of Guid and ApplicationIds of string.

Curious: any particular reason not to use Guid for your client_id?

ilmax commented:
@xperiandri Are you still working on this?

xperiandri commented:
Will do an update based on latest changes and remarks next week

ilmax commented:
Nice!

xperiandri committed:
Separated Application entity primary key from application client identifier

xperiandri committed:
Changed model building

Added required restrictions for appropriate properties.
Added index and alternate key for ApplicationId column of Application.
Renamed Identity tables to be named the same as OpenIddict tables.
Renamed OpenIddictContext to OpenIddictDbContext to emphasis that it is related to Entity Framework as it is done in Identity with IdentityDbContext.

Updated XML documentation to correspond to Identity style.

xperiandri committed:
Updated XML documentation to correspond to Identity style for all the other entities and stores

ilmax commented on src/OpenIddict.EntityFramework/Models/OpenIddictApplication.cs:
```diff
@@ -19,54 +20,57 @@ public class OpenIddictApplication : OpenIddictApplication<string, OpenIddictTok
+        /// Gets or sets the unique identifier for this application.
         /// </summary>
-        public virtual TKey Id { get; set; }
+        public virtual string ApplicationId { get; set; }
```
I prefer call this `ClientId`

xperiandri replied:
What is the idea behind this?
To correspond to `client_id` query parameter of the URL?

ilmax replied:
This is the client identifier called `client_id` in the specs

ilmax commented on src/OpenIddict.EntityFramework/Stores/OpenIddictApplicationStore.cs:
```diff
@@ -60,19 +60,19 @@ public class OpenIddictApplicationStore<TApplication, TToken, TContext, TKey> :
         /// whose result returns the client application corresponding to the identifier.
         /// </returns>
-        public virtual Task<TApplication> FindByIdAsync(string identifier, CancellationToken cancellationToken) {
+        public virtual Task<TApplication> FindByIdAsync(string id, CancellationToken cancellationToken) {
```
I'm pretty sure this method now need to search an application by `ApplicationId` and not by `Id`

xperiandri replied:
Added one more method for that purpose.
Kept this one for style compatibility with Identity reason.

ilmax replied:
Ok, but all the callers should be updated to call the new method (we should lookup a client by `client_id` and not by PK)

xperiandri commented on src/OpenIddict.EntityFramework/Models/OpenIddictApplication.cs:
```diff
@@ -19,54 +20,57 @@ public class OpenIddictApplication : OpenIddictApplication<string, OpenIddictTok
-        /// associated with the current application.
+        /// Gets or sets the callback URL for this application.
         /// </summary>
         public virtual string RedirectUri { get; set; }
```
Maybe naming this as LoginRedirectUri would be more consistent?

xperiandri committed:
Renamed ApplicationId property of Application to ClientId to correspomd to client_id name in specification

Also introduced missing FindByClientIdAsync method in ApplicationStore.
And fixed calls that retrieve application by its client_id. So that they call FindByClientIdAsync instead of FindByIdAsync.

kevinchalet commented:
Merged, thanks :+1: 

https://github.com/openiddict/openiddict-core/commit/4a3b84ddb5a9a0a126ed9427dfd8e071a8b3de81

ilmax commented:
:tada:
